src/AutoMapper/Execution/ExpressionBuilder.cs

```diff
@@ -75,45 +74,22 @@ Expression ClearDestinationCollection()
                     Assign(destinationVariable,
                         ExpressionFactory.ToType(destinationParameter, destinationCollectionType)),
                     destinationVariable.IfNullElse(Empty(), clear),
-                    defaultDestination);
+                    destination);
             }
         }
 
-        private static Expression DefaultDestination(Type destinationType, ProfileMap profileMap)
+        private static Expression DefaultDestination(Type destinationType, Type declaredDestinationType, ProfileMap profileMap)
         {
-            var defaultValue = Default(destinationType);
-            if (profileMap.AllowNullCollections || destinationType == typeof(string) ||
-                !destinationType.IsEnumerableType())
-                return defaultValue;
-            if (destinationType.IsArray)
+            if(profileMap.AllowNullCollections || destinationType == typeof(string) || !destinationType.IsEnumerableType())
             {
-                var destinationElementType = destinationType.GetElementType();
-                return NewArrayBounds(destinationElementType,
-                    Enumerable.Repeat(Constant(0), destinationType.GetArrayRank()));
+                return Default(declaredDestinationType);
             }
-            if (destinationType.IsDictionaryType())
```

Comment:
This is mostly unrelated but it's something that was bothering me. It makes sense to move the null check for the source here, but the mappers know about the destination types. It's messy to repeat all that here.